src/lib/gadgets/elliptic-curve.ts

```diff
@@ -241,16 +271,61 @@ function multiScalarMul(
     getPointTable(Curve, P, windowSizes[i], tableConfigs[i]?.multiples)
   );
 
+  let maxBits = Curve.Scalar.sizeInBits;
+
+  if (useGlv) {
+    maxBits = Curve.Endo.decomposeMaxBits;
+    assert(maxBits < l2, 'decomposed scalars have to be < 2*88 bits');
+
+    // decompose scalars and handle signs
+    let n2 = 2 * n;
+    let scalars2: Field3[] = Array(n2);
+    let points2: Point[] = Array(n2);
+    let windowSizes2: number[] = Array(n2);
+    let tables2: Point[][] = Array(n2);
+    let mrcStack: Field[] = [];
+
+    for (let i = 0; i < n; i++) {
+      let [s0, s1] = decomposeNoRangeCheck(Curve, scalars[i]);
+      scalars2[2 * i] = s0.abs;
+      scalars2[2 * i + 1] = s1.abs;
+
+      let table = tables[i];
+      let endoTable = table.map((P, i) => {
+        if (i === 0) return P;
+        let [phiP, betaXBound] = endomorphism(Curve, P);
+        mrcStack.push(betaXBound);
+        return phiP;
+      });
+      tables2[2 * i] = table.map((P) =>
+        negateIf(s0.isNegative, P, Curve.modulus)
+      );
+      tables2[2 * i + 1] = endoTable.map((P) =>
+        negateIf(s1.isNegative, P, Curve.modulus)
+      );
+      points2[2 * i] = tables2[2 * i][1];
+      points2[2 * i + 1] = tables2[2 * i + 1][1];
+
+      windowSizes2[2 * i] = windowSizes2[2 * i + 1] = windowSizes[i];
+    }
+    reduceMrcStack(mrcStack);
+    // from now on, everything is the same as if these were the original points and scalars
+    points = points2;
+    tables = tables2;
+    scalars = scalars2;
+    windowSizes = windowSizes2;
+    n = n2;
+  }
+
   // slice scalars
-  let b = Curve.order.toString(2).length;
   let scalarChunks = scalars.map((s, i) =>
-    slice(s, { maxBits: b, chunkSize: windowSizes[i] })
+    slice(s, { maxBits, chunkSize: windowSizes[i] })
   );
 
   ia ??= initialAggregator(Curve);
```

Comment:
it could use the endomorphism for the out of circuit scalar multiplication it does when the curve has a cofactor. seems like a very minor improvement though